### ISSUE TITLE
[sklearn#34618] Perf: `PCA(n_components="mle")` uses O(d^3) version of Minka's algorithm when O(d^2) exists

### ISSUE THREAD

**Issue by godofecht** [flooooooooooow/flow#309]
**Mirrored from [scikit-learn#34618](https://github.com/scikit-learn/scikit-learn/issues/34618)**

Original title: Perf: `PCA(n_components="mle")` uses O(d^3) version of Minka's algorithm when O(d^2) exists
Original labels: New Feature, Needs Triage
Created: 2026-07-31T18:16:16Z

---

> [!WARNING]
> This issue is not yet ready for a PR. If you are interested in contributing to scikit-learn, please have a look at our [contributing guidelines](https://scikit-learn.org/dev/developers/contributing.html), and in particular the sections for [new contributors](https://scikit-learn.org/dev/developers/contributing.html#new-contributors) and the ["Needs triage"](https://scikit-learn.org/dev/developers/contributing.html#issues-tagged-needs-triage) label.

### Describe the workflow you want to enable

`PCA(n_components="mle")` is very slow on large datasets since it runs in O(d^3) where `d=min(n_features, n_samples)` and has considerable Python loop overhead.

T Minka has an O(d^2) matlab implementation that incidentally would enable much better vectorization:
https://github.com/tminka/tminka.github.io/blob/master/papers/pca/laplace_pca.zip (see `laplace_pca.m` in that zip)

### Describe your proposed solution

We should port that to Python, or at least the core `pa` reformulation that makes each rank evaluation linear time instead of quadratic. There are a few more tweaks in there, some of which were added years later, that are probably good ideas but would change the behavior of the function slightly. I propose we just port the runtime improvement without changing behavior; future work can look into adapting some of the other improvements.

### Additional context

I'll put together a PR for this.

---

## Flow context

This issue is mirrored from scikit-learn to explore whether Flow encounters the same problem or can approach it differently. Add analysis below.

**Comment by godofecht** [comment]
Moved to https://github.com/godofecht/flow-scikit/issues/114 in the flow-scikit repository. Further library-specific triage and implementation will happen there.
